### PR TITLE
stopped per-instance configs from nuking global defaults

### PR DESCRIPTION
**Pull request by mjksmith**
Fixed a bug where your entire `monitoring.yaml` config changes would be blown away if an instance had any monitoring configs. Probably applies to non-monitoring config files as well.

### CODE REVIEW

**Comment by EvanKrall** [comment]
Probably worth some tests on your deep_merge_dictionaries function

**Comment by solarkennedy** [comment]
Hmm. Don't we do (very verbose) monitoring getters that look in both places? Maybe those are all going to be obsolete with this change. I can't think of other config places that we have dictionaries like that, autoscaling might be a new one though.

**Comment by EvanKrall** [comment]
Context: https://irclogs.yelpcorp.com/view/paasta/2016-03-16.html#161/h161,163,164,165,166,167,183,184,185,186,187,188,189,190,198,199,201,204,205,222,223,224,225,226,227

File: paasta_tools/utils.py [hunk 1]
```diff
@@ -1340,3 +1340,16 @@ def format_table(rows, min_spacing=2):
             expanded_rows.append(expanded_row)
 
     return [(' ' * min_spacing).join(r) for r in expanded_rows]
+
+
+def deep_merge_dictionaries(source, destination):
```
Comment (Rob-Johnson) [inline paasta_tools/utils.py]: As the comment mentions [here](https://stackoverflow.com/questions/20656135/python-deep-merge-dictionary-data) what happens if an element in A is a dict but in B is not?

Comment (mjksmith) [inline paasta_tools/utils.py]: @Rob-Johnson either an `AttributeError` or a `KeyError` will be raised. Most other implementations I've seen just check for this and raise an exception. I suppose in this case we'll want to default to the per-instance configs.

Comment (mjksmith) [inline paasta_tools/utils.py]: This is fixed in the latest commit. If a value and a dictionary share the same key, whatever is in the per-instance config will override the per-service config.

**Comment by Rob-Johnson** [comment]
How would you feel about this creating a new dictionary,  rather than modifying the dict passed in?

```
foo = {foo: {bar: baz}}; 
blargh = {foo: { bar: otherbaz}}
merged = deep_merge_dictionaries(bar, foo)
assert foo == {foo: {bar: baz}}
```

**Comment by mjksmith** [comment]
@Rob-Johnson I agree with this being better. Changed in dd9b421

**Comment by mjksmith** [comment]
@solarkennedy looking at those monitoring getters in `paasta_tools.monitoring_tools`, they're only used to get specific keys, so they're a lot faster than merging entire dictionaries.

**Comment by Rob-Johnson** [comment]
shipit

**Pull request merged**